gateway/hostname.go

```diff
@@ -306,22 +306,59 @@ func isHTTPSRequest(r *http.Request) bool {
 
 // Converts a FQDN to DNS-safe representation that fits in 63 characters:
 // my.v-long.example.com → my-v--long-example-com
-func toDNSLinkDNSLabel(fqdn string) (dnsLabel string, err error) {
+// Implements specification from https://specs.ipfs.tech/http-gateways/subdomain-gateway/#host-request-header
+func InlineDNSLink(fqdn string) (dnsLabel string, err error) {
+	/* What follows is an optimized version this three-liner:
 	dnsLabel = strings.ReplaceAll(fqdn, "-", "--")
 	dnsLabel = strings.ReplaceAll(dnsLabel, ".", "-")
 	if len(dnsLabel) > dnsLabelMaxLength {
 		return "", fmt.Errorf("DNSLink representation incompatible with DNS label length limit of 63: %s", dnsLabel)
 	}
 	return dnsLabel, nil
+	*/
+	result := make([]byte, 0, len(fqdn))
+	for i := 0; i < len(fqdn); i++ {
+		char := fqdn[i]
+		if char == '-' {
+			result = append(result, '-', '-')
+		} else if char == '.' {
+			result = append(result, '-')
+		} else {
+			result = append(result, char)
+		}
+	}
+	if len(result) > dnsLabelMaxLength {
+		return "", fmt.Errorf("inlined DNSLink incompatible with DNS label length limit of 63: %q", result)
+	}
+	return string(result), nil
 }
 
 // Converts a DNS-safe representation of DNSLink FQDN to real FQDN:
 // my-v--long-example-com → my.v-long.example.com
-func toDNSLinkFQDN(dnsLabel string) (fqdn string) {
+// Implements specification from https://specs.ipfs.tech/http-gateways/subdomain-gateway/#host-request-header
+func UninlineDNSLink(dnsLabel string) (fqdn string) {
```

Comment:
```suggestion
// UninlineDNSLink implements specification from https://specs.ipfs.tech/http-gateways/subdomain-gateway/#host-request-header
func UninlineDNSLink(dnsLabel string) (fqdn string) {
```

Just Go convention-stuff.